lib/src/main/java/com/dynatrace/file/util/PollBasedFilePoller.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright 2022 Dynatrace LLC
+ *
+ * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * <p>http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.dynatrace.file.util;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Logger;
+
+class PollBasedFilePoller extends AbstractFilePoller {
+  private final AtomicBoolean changedSinceLastPoll = new AtomicBoolean(false);
+  private static final Logger logger = Logger.getLogger(PollBasedFilePoller.class.getName());
+  private volatile long lastUpdatedAt = 0;
+
+  protected PollBasedFilePoller(Path filePath, Duration pollInterval) {
+    super(filePath);
+    if (pollInterval == null) {
+      throw new IllegalArgumentException("Poll interval cannot be null");
+    }
+
+    ScheduledExecutorService scheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactory() {
+              @Override
+              public Thread newThread(Runnable r) {
+                Thread t = new Thread(null, r, "PollBasedFilePoller", 0);
+                t.setDaemon(true);
+                return t;
+              }
+            });
+    logger.finer(() -> String.format("Polling every %dms", pollInterval.toMillis()));
+    scheduledExecutorService.scheduleAtFixedRate(
+        this::poll, pollInterval.toNanos(), pollInterval.toNanos(), TimeUnit.NANOSECONDS);
+
+    // poll once initially and synchronously to make sure that the atomic data stores are
+    // initialized before the end of the constructor.
+    poll();
+  }
+
+  @Override
+  public boolean fileContentsUpdated() {
+    // get the current value and reset to false
+    return changedSinceLastPoll.getAndSet(false);
+  }
+
+  private void poll() {
+    synchronized (this) {
+      try {
+        long modifiedAt = Files.getLastModifiedTime(absoluteFilePath).toMillis();
+        long previouslyModifiedAt = lastUpdatedAt;
+        lastUpdatedAt = modifiedAt;
+
+        if (modifiedAt >= previouslyModifiedAt) {
```

Comment:
```suggestion
        long previouslyModifiedAt = lastUpdatedAt;
        lastUpdatedAt = Files.getLastModifiedTime(absoluteFilePath).toMillis();

        if (lastUpdatedAt >= previouslyModifiedAt) {
```
This should do the same thing within the `synchronized`, right?

Reply:
(Could also consider renaming it to `prevLastModifiedAt` and `lastModifiedAt` to indicate it's about the same file attribute, otherwise one might think there's a difference between modified and updated.)

Reply:
Simplified it by moving the assignment below the if block. I think it should do the same, however, I am not sure if it has to query to main memory every time since it's `volatile`, that's why it might be a good idea to keep it in a separate variable inside the synchronized block